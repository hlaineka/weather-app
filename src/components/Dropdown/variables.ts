export type Option = {
  value: string;
  label: string;
  lat: number;
  lng: number;
};

export const options: Array<Option> = [
  {
    value: 'all',
    label: 'Kaikki kaupungit',
    lat: 0,
    lng: 0,
  },
  {
    value: 'espoo',
    label: 'Espoo',
    lat: 60.25,
    lng: 24.6667,
  },
  {
    value: 'jyvaskyla',
    label: 'Jyväskylä',
    lat: 62.2415,
    lng: 25.7209,
  },
  {
    value: 'kuopio',
    label: 'Kuopio',
    lat: 62.8924,
    lng: 27.677,
  },
  {
    value: 'tampere',
    label: 'Tampere',
    lat: 61.4991,
    lng: 23.7871,
  },
];
