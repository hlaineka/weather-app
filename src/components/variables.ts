export type Color = keyof typeof palette;

export const palette = {
  blue: '#00A5E5',
  blueBackground: '#E5F6FD',
  white: '#FFFFFF',
  background: '#F8F9FA',
  black: '#262626',
  greyText: '#70757A',
  greyBorder: '#E6E6E6',
};

export const SPACING = '15px';
export const CONTENT_WIDTH = `calc(100vw - (2 * ${SPACING}))`;
export const MAX_CONTENT_WIDTH = '550px';
