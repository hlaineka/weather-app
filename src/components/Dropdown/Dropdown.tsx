import React, { useState } from 'react';
import styled from 'styled-components';
import { Chevron } from './Chevron';
import { options, Option } from './variables';
import { CONTENT_WIDTH, palette } from '../variables';

type Props = {
  setDropdownSelection: (selected: Option) => void;
  dropdownSelection: Option;
};

export const Dropdown: React.FC<Props> = ({
  setDropdownSelection,
  dropdownSelection,
}) => {
  const [isDropdownMenuVisible, setDropdownMenuVisible] = useState(false);

  const toggleDropdownMenu = () => {
    setDropdownMenuVisible(!isDropdownMenuVisible);
  };

  const namespace = 'cities';
  const value = dropdownSelection;

  return (
    <DropdownContainer>
      <DropdownButton
        onClick={toggleDropdownMenu}
        role="combobox"
        aria-haspopup="listbox"
        aria-controls={`${namespace}_dropdown`}
        aria-labelledby={`${namespace}_label`}
        aria-expanded={isDropdownMenuVisible}
        aria-activedescendant={`${namespace}_element_${value.value}`}
      >
        {value.label}
        {isDropdownMenuVisible ? (
          <Chevron variant="up" color="black" />
        ) : (
          <Chevron variant="down" color="black" />
        )}
      </DropdownButton>
      {isDropdownMenuVisible && (
        <DropdownMenu>
          <ul
            className="select-dropdown"
            role="listbox"
            id={`${namespace}_dropdown`}
            tabIndex={-1}
            aria-multiselectable={false}
          >
            {options.map(
              (option: Option) =>
                option &&
                option.label && (
                  <li
                    key={option.value}
                    className={`${namespace}_element_${option.value}`}
                    id={`${namespace}_element_${option.value}`}
                    aria-selected={option.value === value.value}
                    role="option"
                    onClick={() => {
                      setDropdownSelection(option);
                      toggleDropdownMenu();
                    }}
                  >
                    <label key={option.value}>{option.label}</label>
                  </li>
                ),
            )}
          </ul>
        </DropdownMenu>
      )}
    </DropdownContainer>
  );
};

const DropdownContainer = styled.div`
  display: inline-block;
  height: fit-content;
  position: relative;
  width: 100%;
`;

const DropdownMenu = styled.div`
  background-color: white;
  border: 1px solid ${palette.greyBorder};
  box-sizing: border-box;
  max-width: ${CONTENT_WIDTH};
  padding-bottom: 8pt;
  padding-top: 10pt;
  position: absolute;
  right: 0;
  width: 100%;
  z-index: 1;

  ul {
    list-style-type: none;
    margin: 0;
    max-width: ${CONTENT_WIDTH};
    padding: 0;

    li {
      margin-bottom: 2pt;
      max-width: ${CONTENT_WIDTH};
      padding-left: 10pt;
      padding-right: 10pt;

      &[aria-selected='true'] {
        background: ${palette.blueBackground};
      }
    }
  }
`;

const DropdownButton = styled.button`
  align-items: center;
  background-color: white;
  border: 1px solid ${palette.greyBorder};
  box-sizing: border-box;
  color: inherit;
  display: flex;
  flex-direction: row;
  font-size: 13pt;
  justify-content: space-between;
  line-height: 13pt;
  max-width: ${CONTENT_WIDTH};
  padding: 10pt;
  text-align: left;
  width: 100%;

  &:hover + & {
    display: inline-block;
  }
`;
