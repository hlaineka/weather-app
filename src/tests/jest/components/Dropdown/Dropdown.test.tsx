import React from 'react';
import renderer, { act } from 'react-test-renderer';
import Dropdown from '../../../../components/Dropdown';
import { Option } from '../../../../components/Dropdown/variables';

describe('Dropdown:', () => {
  describe('Rendering:', () => {
    test('should render closed dropdown menu properly', () => {
      //given
      let dummySelected = {
        value: 'tampere',
        label: 'Tampere',
        lat: 0,
        lng: 0,
      };
      const dummyHandlesetDropdownSelection = (selected: Option) => {
        dummySelected = selected;
      };
      const dummyComponentProps = {
        setDropdownSelection: dummyHandlesetDropdownSelection,
        dropdownSelection: dummySelected,
      };

      // when
      const componentRenderer = renderer.create(
        <Dropdown {...dummyComponentProps} />,
      );
      const tree = componentRenderer.toJSON();

      // then
      expect(tree).toMatchSnapshot();
    });
  });
  test('should render open dropdown menu properly', () => {
    //given
    let dummySelected = {
      value: 'tampere',
      label: 'Tampere',
      lat: 0,
      lng: 0,
    };
    const dummyHandlesetDropdownSelection = (selected: Option) => {
      dummySelected = selected;
    };
    const dummyComponentProps = {
      setDropdownSelection: dummyHandlesetDropdownSelection,
      dropdownSelection: dummySelected,
    };

    // when
    const componentRenderer = renderer.create(
      <Dropdown {...dummyComponentProps} />,
    );

    act(() => {
      componentRenderer.root.findByType('button').props.onClick();
    });
    const tree = componentRenderer.toJSON();

    // then
    expect(tree).toMatchSnapshot();
  });

  describe('Selecting:', () => {
    test('should select Espoo properly', async () => {
      //given
      let dummySelected = {
        value: 'tampere',
        label: 'Tampere',
        lat: 0,
        lng: 0,
      };
      const dummyHandlesetDropdownSelection = (selected: Option) => {
        dummySelected = selected;
      };
      const dummyComponentProps = {
        setDropdownSelection: dummyHandlesetDropdownSelection,
        dropdownSelection: dummySelected,
      };

      // when
      const componentRenderer = renderer.create(
        <Dropdown {...dummyComponentProps} />,
      );

      await act(async () => {
        componentRenderer.root.findByType('button').props.onClick();
      });

      await act(async () => {
        componentRenderer.root
          .findByProps({ className: 'cities_element_espoo' })
          .props.onClick();
      });

      const tree = componentRenderer.toJSON();

      // then
      expect(tree).toMatchSnapshot();
      expect(dummySelected.label).toBe('Espoo');
    });
  });
});
