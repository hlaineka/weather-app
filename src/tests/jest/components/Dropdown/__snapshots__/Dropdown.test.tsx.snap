// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown: Rendering: should render closed dropdown menu properly 1`] = `
<div
  className="sc-bczRLJ RcPSw"
>
  <button
    aria-activedescendant="cities_element_tampere"
    aria-controls="cities_dropdown"
    aria-expanded={false}
    aria-haspopup="listbox"
    aria-labelledby="cities_label"
    className="sc-dkzDqf hfXXFQ"
    onClick={[Function]}
    role="combobox"
  >
    Tampere
    <svg
      fill="none"
      height={14}
      viewBox="0 0 10 6"
      width={9}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0.292893 0.292893C0.683418 -0.0976314 1.31658 -0.0976314 1.70711 0.292893L5 3.58579L8.29289 0.292893C8.68342 -0.0976314 9.31658 -0.0976314 9.70711 0.292893C10.0976 0.683417 10.0976 1.31658 9.70711 1.70711L5.70711 5.70711C5.51957 5.89464 5.26522 6 5 6C4.73478 6 4.48043 5.89464 4.29289 5.70711L0.292893 1.70711C-0.097631 1.31658 -0.097631 0.683417 0.292893 0.292893Z"
        fill="#262626"
      />
    </svg>
  </button>
</div>
`;

exports[`Dropdown: Selecting: should select Espoo properly 1`] = `
<div
  className="sc-bczRLJ RcPSw"
>
  <button
    aria-activedescendant="cities_element_tampere"
    aria-controls="cities_dropdown"
    aria-expanded={false}
    aria-haspopup="listbox"
    aria-labelledby="cities_label"
    className="sc-dkzDqf hfXXFQ"
    onClick={[Function]}
    role="combobox"
  >
    Tampere
    <svg
      fill="none"
      height={14}
      viewBox="0 0 10 6"
      width={9}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0.292893 0.292893C0.683418 -0.0976314 1.31658 -0.0976314 1.70711 0.292893L5 3.58579L8.29289 0.292893C8.68342 -0.0976314 9.31658 -0.0976314 9.70711 0.292893C10.0976 0.683417 10.0976 1.31658 9.70711 1.70711L5.70711 5.70711C5.51957 5.89464 5.26522 6 5 6C4.73478 6 4.48043 5.89464 4.29289 5.70711L0.292893 1.70711C-0.097631 1.31658 -0.097631 0.683417 0.292893 0.292893Z"
        fill="#262626"
      />
    </svg>
  </button>
</div>
`;

exports[`Dropdown: should render open dropdown menu properly 1`] = `
<div
  className="sc-bczRLJ RcPSw"
>
  <button
    aria-activedescendant="cities_element_tampere"
    aria-controls="cities_dropdown"
    aria-expanded={true}
    aria-haspopup="listbox"
    aria-labelledby="cities_label"
    className="sc-dkzDqf hfXXFQ"
    onClick={[Function]}
    role="combobox"
  >
    Tampere
    <svg
      fill="none"
      height={14}
      viewBox="0 0 10 6"
      width={9}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M9.70711 5.70708C9.31658 6.0976 8.68342 6.0976 8.29289 5.70708L5 2.41418L1.70711 5.70708C1.31658 6.0976 0.683417 6.0976 0.292892 5.70708C-0.097632 5.31655 -0.0976319 4.68339 0.292892 4.29286L4.29289 0.292862C4.48043 0.105326 4.73478 -3.10365e-05 5 -3.10103e-05C5.26522 -3.09842e-05 5.51957 0.105326 5.70711 0.292862L9.70711 4.29286C10.0976 4.68339 10.0976 5.31655 9.70711 5.70708Z"
        fill="#262626"
      />
    </svg>
  </button>
  <div
    className="sc-gsnTZi iuBPRX"
  >
    <ul
      aria-multiselectable={false}
      className="select-dropdown"
      id="cities_dropdown"
      role="listbox"
      tabIndex={-1}
    >
      <li
        aria-selected={false}
        className="cities_element_all"
        id="cities_element_all"
        onClick={[Function]}
        role="option"
      >
        <label>
          Kaikki kaupungit
        </label>
      </li>
      <li
        aria-selected={false}
        className="cities_element_espoo"
        id="cities_element_espoo"
        onClick={[Function]}
        role="option"
      >
        <label>
          Espoo
        </label>
      </li>
      <li
        aria-selected={false}
        className="cities_element_jyvaskyla"
        id="cities_element_jyvaskyla"
        onClick={[Function]}
        role="option"
      >
        <label>
          Jyväskylä
        </label>
      </li>
      <li
        aria-selected={false}
        className="cities_element_kuopio"
        id="cities_element_kuopio"
        onClick={[Function]}
        role="option"
      >
        <label>
          Kuopio
        </label>
      </li>
      <li
        aria-selected={true}
        className="cities_element_tampere"
        id="cities_element_tampere"
        onClick={[Function]}
        role="option"
      >
        <label>
          Tampere
        </label>
      </li>
    </ul>
  </div>
</div>
`;
